### ISSUE TITLE
Exceptions that escape SpringMVC DispatcherServlet should be logged with trace & span IDs

### ISSUE THREAD

**Issue by jfuerth** [spring-cloud/spring-cloud-sleuth#714]
Hi,

We are re-raising the issue reported in #579 and #542 because we've tried the suggested solution of handling and logging all the errors ourselves, but we have yet to land on a workable solution.

We tried adding a method like this to our global ControllerAdvice bean:

```
    @ExceptionHandler(Exception.class)
    ResponseEntity<ErrorPayload> otherwiseUnhandledException(Exception e) {
        log.error("Unexpected exception occurred. Will return HTTP 500.", e);
        return new ResponseEntity<>(new ErrorPayload(e.getMessage(), ""), INTERNAL_SERVER_ERROR);
    }
```

The error is logged with the correct trace and span IDs (great!) but the killer side effect is that all the internal Spring MVC exceptions that do nice things like return 400 on missing mandatory by our ExceptionHandler method and thus return HTTP 500 to our clients.

We also experimented with registering our own `BasicErrorController` bean based on the callout box in the docs linked from #579 (https://docs.spring.io/spring-boot/docs/current-SNAPSHOT/reference/htmlsingle/#boot-features-error-handling). We see that it is getting instantiated, but its `onError` method is not getting called for our failing JSON API requests:

```
@Slf4j
@Controller
public class MyErrorController extends BasicErrorController {

    public MyErrorController(ErrorAttributes errorAttributes) {
        super(errorAttributes, new ErrorProperties());
        log.info("Created error controller with errorAttributes {}", errorAttributes);
    }

    @RequestMapping(produces = "application/json")
    public ErrorPayload onError() {
        log.error("KABOOM!");
        return new ErrorPayload("Something broke (TODO: resolve actual exception)", "????");
    }
}
```

We'd really appreciate any advice you can offer on how to preserve Spring MVC's default response codes while also including the Sleuth trace and span IDs in the MDC.

Thanks!

**Comment by marcingrzejszczak** [comment]
Like presented in the commit message:

> without this change any exception that is not caught by a filter, get caught and logged by server's internals (e.g. Tomcat's Valve). To make the Valve log the exception together with the tracing information we would have to allow the tracing context remain in the thread after TraceFilter gets executed. That is problematic cause we're polluting the ThreadLocal and would have to assume that some component will eventually clear the context.
>
> with this change we're trying to solve the issue from a different angle. Whenever an uncaught exception is thrown, we are already catching it in the `catch(...) {}` clause. It's enough to just log it at the error level and that way, regardless of the underlying server implementation (Tomcat, Undertow) we will log the uncaught exception and rethrow it.

**Comment by brenuart** [comment]
>  we would have to allow the tracing context remain in the thread after TraceFilter gets executed. That is problematic cause we're polluting the ThreadLocal and would have to assume that some component will eventually clear the context.

Why not using a `javax.servlet.ServletRequestListener` to listen for the completion of the request and clear the ThreadLocal? Request completed event occurs at the very end - including after it is forwarded to the /error page in case of exception...

In a Tomcat environment, the current approach causes the exception to be logged twice: once by the TraceFilter then by the Tomcat valve.

**Comment by marcingrzejszczak** [comment]
But the Tomcat Valve is not traced isn't it? If that's not the case, can you show an example of using the `javax.servlet.ServletRequestListener` ?

**Comment by brenuart** [comment]
> But the Tomcat Valve is not traced isn't it?

Not sure I understand your question... If that's what you mean, I confirm the exception stack trace logged by Tomcat is *NOT* traced (reason why you decided to log it first inside the TraceFilter).

> can you show an example of using the `javax.servlet.ServletRequestListener` ?

I had a look at the TraceFilter code and honestly it looks quite complex and I still don't understand why and when a Span must be started/closed/detached/suspended/etc.

It's your comment saying you would have to leave stuff on the ThreadLocal until the end of the request but don't know how to clean it after that made me think about the ServletRequestListener. I posted to share ideas...

I don't have enough Sleuth knowledge for this task I'm afraid - but I would be glad to help if you give me some more explanation about the TraceFilter internals!

**Comment by marcingrzejszczak** [comment]
> Not sure I understand your question... If that's what you mean, I confirm the exception stack trace logged by Tomcat is NOT traced

Yes, you answered my question. 

> (reason why you decided to log it first inside the TraceFilter).

Yes, that's why this problem is solved and I see no reason to do anything inside Tomcat Valve. With the current approach we instrument any uncaught exceptions regardless whether they are Tomcat or not.

> I had a look at the TraceFilter code and honestly it looks quite complex and I still don't understand why and when a Span must be started/closed/detached/suspended/etc - and basically how it works.

Due to the fact that the filter is used in different dispatches, with & without Error controller / Method Handler / Zuul filter it's complex - I agree.

> It's your comment saying you would have to leave stuff on the ThreadLocal until the end of the request but don't know how to clean it after that made me think about the ServletRequestListener. I posted to share ideas...

I don't understand the problem. Do we even have one? :P

**Comment by brenuart** [comment]
> I don't understand the problem. Do we even have one?

The "problem" is the exception is now logged *twice* instead of only once.

>  I see no reason to do anything inside Tomcat Valve

There is nothing todo in the Tomcat Valve... The proposition was to rely on a `ServletRequestListener` to delay the ThreadLocal cleanup (or Span close if that is the same) until the request is fully processed. The benefit is this listener encompass the entire processing of the request, including all forwards that may happen (like /error) whereas the ServletFilter is possibly invoked multiple times (adding to its "complexity").

Do you want me to create a separate issue for this "double exception logging" stuff?

**Comment by brenuart** [comment]
Another potential issue I'm thinking of is the case of the `ClientAbortException` thrown by Tomcat when the connection is closed by the client. This exception can happen from within Tomcat itself during the flush of its response buffer at the end of request processing but also from within the application when the response is too large to fit in the buffer. It is not meant to be handled by the application and is therefore a candidate for an "uncatched exception" and is likely to be logged by the TraceFilter too.
Tomcat on the other hand does NOT log this exception because it is considered a "normal" case.

**Comment by marcingrzejszczak** [comment]
> Do you want me to create a separate issue for this "double exception logging" stuff?

I don't agree that this is a problem. We're catching the exception and logging it at a higher level of abstraction.

> Tomcat on the other hand does NOT log this exception because it considers it a "normal" case. I have to double check, but if I remember correctly, it won't even forward the request to the error endpoint (does it break assumptions made by the TraceFilter?)...

Please feel free to file another issue and create a sample of what you suggest to be done, maybe that way it will be easier to understand the whole problem.
